Allow importing package.json

iamdoron opened the pull request:
Hi,

Basically, allow to import `package.json`. 
For example, from `src/App.js`:
```javascript
import { version } from '../package.json'
```
Only the app own `package.json` is allowed to be imported.

It closes https://github.com/facebookincubator/create-react-app/issues/2466


<img width="457" alt="import_packagejson" src="https://cloud.githubusercontent.com/assets/3956042/26762761/41a01a72-4950-11e7-838e-c7afb9b0b1ea.png">

rogovdm commented:
It it a safe practice to do so?

iamdoron commented:
@rogovdm, I guess it depends what kind of date is in the `package.json`. Also, in terms of bundle size it might not be efficient to import the whole `package.json`.

Maybe it's better to only include the version property from the `package.json`

evenchange4 commented:
I have done the similar use case by adding `REACT_APP_VERSION` to build script:

```diff
+ "build": "REACT_APP_VERSION=$(node -pe 'require(\"./package.json\").version') react-scripts build",
```

Also, you can set it to a constant while developing/testing.

```
// .env.development

REACT_APP_VERSION=0.0.0
```

[source code](https://github.com/MCS-Lite/mcs-lite/blob/master/packages/mcs-lite-mobile-web/package.json#L26)

## Screenshot
<img width="270" alt="_2017-06-05_11_16_56" src="https://cloud.githubusercontent.com/assets/1527371/26769484/ee6da486-49e1-11e7-999d-4361a14aa2f9.png">

iamdoron commented on packages/react-dev-utils/ModuleScopePlugin.js:
```diff
@@ -37,24 +37,28 @@ class ModuleScopePlugin {
+      const requestRelativeToRoot = path.relative(
+        request.descriptionFileRoot,
+        requestFullPath
+      );
```
@gaearon you are totally right. My previous code wasn't working . I'm not sure what I did there, but this is what I meant to do - to check this is the package.json that's in the root. I didn't want to assume package.json always lives in `../src` in case the src is more than one level deep like `true/src`

gaearon commented:
I don't understand newly pushed commits. I was referring to using existing variable (which has the right thing). You renamed an old variable but it still calculates the wrong thing.

iamdoron commented:
I think it's OK, but maybe I'm missing something
there are two types or "relatives". 
relative to src
relative to root

I wanted to check this is the package.json that resides in root:
```javascript
const requestRelativeToRoot = path.relative(
        request.descriptionFileRoot,
        requestFullPath
);
```

gaearon commented:
I'm not sure myself. Just make sure importing `../somethingelse.json` is still an error.

iamdoron commented:
now when I run it I see:

`require("../../package_.json")`
```
Module not found: You attempted to import ../../package_.json
```

`require("../package_.json")`
```
Module not found: You attempted to import ../package.json
```

and `require("../../package.json")` works because it's inside the root

Timer commented:
Wouldn't it be easier to simply use [this variable](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-dev-utils/ModuleScopePlugin.js#L47) and add:
```js
      if (requestRelative === `..${path.sep}package.json`) {
        return callback();
      }
```

iamdoron commented:
@Timer  I wanted it to check from the root and not `src/../package.json` in case the source folder is two levels or more from the root
but maybe it's enough to assume that `src/../package.json` is the right one

Timer commented:
Good point, @iamdoron.
Maybe a regex pattern like this?
```js
/^(..[/|\\])+package.json$/
```

iamdoron commented on packages/react-dev-utils/ModuleScopePlugin.js:
```diff
@@ -37,24 +37,21 @@ class ModuleScopePlugin {
+        request.__innerRequest_request
       );
+      const requestRelative = path.relative(appSrc, requestFullPath);
+      if (/^(..[/|\\])+package(.json)?$/.test(requestRelative)) {
```
@Timer I changed it to regex, I added '?' in  `(.json)?`  to make it optional. that way you can `require("../package")`

Timer replied:
It should always include the extension no matter what.

viankakrisna commented:
How about adding excludes option to ModuleScopePlugin?
```
      new ModuleScopePlugin({
        appSrc: paths.appSrc,
        excludes: [paths.appPackageJson]
      }),
```
```
      if (excludes.includes(request.path)){
        return callback()
      }

```

Timer commented:
@viankakrisna that's probably best long-term; what do you think @iamdoron?

gaearon commented:
What's the status on this? What are requested changes?

Timer commented:
We should strictly enforce the `.json` extension so people don't use this as an escape hatch to import `package.js` from the root.

iamdoron commented:
sorry for the delay
I implemented @viankakrisna suggestion

I wanted to include
```js
import { version, name } from '../package.json';
```
in the template App.js, but wasn't so easy 😅
On the e2e test (and for regular users of react-scripts) the package.json is located at `'../package.json'`, but when you run `npm run start` / `npm run build` inside `create-react-app` repo  it was located at `'../../package.json'`
I changed the config/paths.js code such that it will assume the package.json is inside the template (and not react-scripts/package.json). It makes more sense to me, but maybe I missed something.

Timer commented:
Thanks for getting these changes in!
I don't want to advertise this in the template, could you please remove that?

Timer commented on packages/react-scripts/config/paths.js:
```diff
@@ -105,13 +105,13 @@ if (
-    publicUrl: getPublicUrl(resolveOwn('package.json')),
-    servedPath: getServedPath(resolveOwn('package.json')),
+    publicUrl: getPublicUrl(resolveOwn('template/package.json')),
+    servedPath: getServedPath(resolveOwn('template/package.json')),
```
I'm pretty sure these changes are not necessary anymore.

Timer commented on packages/react-scripts/template/package.json:
```diff
@@ -0,0 +1,5 @@
+{
```
We can delete this file, I'd like to keep this PR focused. If we have a good reason for adding this and the corresponding `paths.js` changes, we can propose that in a new PR.

iamdoron replied:
OK, but I think it has no value on it's own (as a PR), it's just theoretically wrong to say that the package.json path is react-scripts/package.json, it should be the template/package.json. Right now it's not an issue but maybe in the future

Timer replied:
We can propose this change as part of enhancing contributor workflow, I could see some potential value. 😄

Timer commented on packages/react-dev-utils/ModuleScopePlugin.js:
```diff
@@ -13,8 +13,9 @@ const chalk = require('chalk');
-  constructor(appSrc) {
+  constructor(appSrc, allowedPaths) {
     this.appSrc = appSrc;
+    this.allowedPaths = new Set(allowedPaths);
```
I assumed this is safe when `allowedPaths` is undefined? We could add a default above to be explicit:
```js
constructor(appSrc, allowedPaths = []) {
```

iamdoron replied:
it is safe, new Set(undefined) is an empty set
but I can add it 👍

Timer commented:
@gaearon how do you feel about this PR versus [this solution](https://github.com/facebookincubator/create-react-app/issues/2466#issuecomment-312016096)? I almost prefer the latter, but I'm not sure if there's other benefits of `package.json`, such as `name`, `author`, etc.

Timer commented on packages/react-dev-utils/ModuleScopePlugin.js:
```diff
@@ -13,8 +13,9 @@ const chalk = require('chalk');
 
 class ModuleScopePlugin {
-  constructor(appSrc) {
+  constructor(appSrc, allowedPaths = []) {
```
Small nit, how about `allowedFiles`? Paths makes it sound like you can give it a directory.

Timer commented on packages/react-dev-utils/README.md:
```diff
@@ -57,7 +57,7 @@ module.exports = {
 
 
-#### `new ModuleScopePlugin(appSrc: string)`
+#### `new ModuleScopePlugin(appSrc: string, allowedPaths: [string])`
```
Ditto, `allowedFiles`. And we should probably let them know it's optional `allowedFiles?: string[]`.

br0p0p commented:
What's the status on this?

Timer commented:
@br0p0p waiting for CI to pass, now. 😄

Timer commented:
Thanks for this, @iamdoron!

Timer commented:
Hi there! This change is out in `react-scripts@1.0.11`; please give it a go! Thanks.

br0p0p commented:
@Timer great, thanks for this!
